### ISSUE TITLE
passwd blocks interrupt signal

### ISSUE THREAD

**Issue by archenemies** [shadow-maint/shadow#319]
Sometimes when I mistype my password, I don't want to wait a full 1000 milliseconds for the "passwd: Authentication failure" message to appear, so I hit Ctrl-C to exit the program. However, it doesn't work, apparently `passwd` is blocking SIGINT. Is there an advantage to doing this? I can't think of any security reason to prevent a user from aborting the `passwd` program via the keyboard. I've been wondering why this is designed this way for 20 years, maybe someone can fill me in or update the manual page with an explanation.

**Comment by edneville** [comment]
I believe that is to avoid interrupting a write sequence that could be stopped with ctrl-c. I don't think it is well suited as a guard though. In theory, removing the delay (if not using PAM) would permit many more guesses at the terminal with something like `expect`.

**Comment by archenemies** [comment]
@edneville Interrupting a write sequence? Can you give more detail? Are you saying that programs which call `write()` need to block Ctrl-C?

**Comment by edneville** [comment]
@archenemies, nope, that's not what I'm saying. Looking through some of the code, I saw the following commentary:

```
/* ... Any keyboard signals are set to be ignored. */
pwd_init ();
```

Likely it was added to reduce possible inconsistency through the suite as it is called in several places. Listening to ctrl-c would be possible by removing the call to pwd_init here.

**Comment by archenemies** [comment]
@edneville Thanks for looking at the code. Is there a reason that `passwd` needs to ignore Ctrl-C, or can we change it to listen to the user?

**Comment by hallyn** [comment]
Well, the comment says

> Call this before modifying password files, so that
> * it is less likely to fail in the middle of operation.

So if you wanted to support ^c, you just need to hook it with a fn that will just stop accepting keyboard entry, but still properly close all the files.

**Comment by archenemies** [comment]
@hallyn Thanks for your reply. I ran `passwd` with a debugger and it looks like the prompt is happening inside the PAM library, in `pam_chauthtok`. Since this function seems to be doing everything from prompting the user to writing system files, it is not possible to just move `pwd_init()` so that it is called after the user interaction.

However, it seems odd that PAM wouldn't contain its own checks to ensure data integrity. I can look at the code for PAM and see if it doesn't handle signal blocking itself.

If you would be willing to look over a PR to fix this at some point, then I can do some more investigation and report back with an outline of what I think should be changed.

**Comment by hallyn** [comment]
> If you would be willing to look over a PR to fix this at some point, then I can do some more investigation and report back with an outline of what I think should be changed.

Sounds great - thank you.

**Comment by hallyn** [comment]
BTW - I suspect the pwd_init() function predates pam integration.  Could be wrong.
